### DISCUSSION TITLE
`urfave/cli` vs https://github.com/jessevdk/go-flags

### DISCUSSION THREAD

**Discussion by abitrolly** [urfave/cli#2006] [General]
I like the declarative way of https://github.com/jessevdk/go-flags to specify flags. It is basically the inverse of my fantasy from https://github.com/urfave/cli/issues/1607

```go
var opts struct {
	// Slice of bool will append 'true' each time the option
	// is encountered (can be set multiple times, like -vvv)
	Verbose []bool `short:"v" long:"verbose" description:"Show verbose debug information"`

	// Example of automatic marshalling to desired type (uint)
	Offset uint `long:"offset" description:"Offset"`

	// Example of a callback, called each time the option is found.
	Call func(string) `short:"c" description:"Call phone number"`

	// Example of a required flag
	Name string `short:"n" long:"name" description:"A name" required:"true"`

	// Example of a flag restricted to a pre-defined set of strings
	Animal string `long:"animal" choice:"cat" choice:"dog"`
...
```

Now I wonder what limitations appear from this approach. Like if there are features of `urfave/cli` that can not be supported by this lib or very hard to implement? And vice versa.

**Comment by dearchap** [comment]
@abitrolly https://github.com/urfave/sflags  . recently forked it as the origin repo hasnt been maintained for past 3-4 years. I've updated the repo in our org with current go version and support for cli v2 and v3 !!!!. You can hack it to your heart's content and let me know what else needs to be added.

**Reply by abitrolly** [reply]
Wow. Great work, @m0sth8 and @furdarius.

**Reply by abitrolly** [reply]
> You can hack it to your heart's content and let me know what else needs to be added.

@dearchap separate examples of using it with each of the lib mentioned. Right now there is only README with `flag` example. For `urfave/cli` it could be nice to see side by side comparison of example without sflags and with it (with binary sizes). In a very ideal world. :D

**Reply by m0sth8** [reply]
That's awesome. @dearchap if you want to maintain the library as part of urfave I can transfer the project. (I don't know how to do this and need some help) .

**Reply by dearchap** [reply]
@m0sth8 Yes we can maintain it. I think I need to add you to the urfave org and then you should be able to transfer it.

**Reply by dearchap** [reply]
@abitrolly Examples are present. https://github.com/urfave/sflags/tree/main/examples   . Here is the one for urfave/cli.v2 https://github.com/urfave/sflags/blob/main/examples/urfave_cli/main.go

This is an ancillary project and not core to urfave/cli. If you want to add metrics/tests to compute binary size of sflags with all the different libraries then please go ahead and submit a PR. Will be glad to review,

**Reply by dearchap** [reply]
@m0sth8 I've added you to https://github.com/orgs/urfave/teams/sflags with admin permissions for urfave/sflags. If you go to the settings in your own org , you can transfer the sflags repo to this org. https://docs.github.com/en/repositories/creating-and-managing-repositories/transferring-a-repository

**Reply by m0sth8** [reply]
@dearchap thank you. I got this error message "urfave/sflags already exists and You don’t have the permission to create public repositories on urfave"
<img width="752" alt="Screenshot 2024-11-11 at 7 47 13 AM" src="https://github.com/user-attachments/assets/f5083c10-ea06-4a9a-a9a9-dfbddecf6e17">

**Reply by dearchap** [reply]
@m0sth8 Did you join the team request I sent ? Also I renamed sflags to sflags-bak in ufave/cli for now

![image](https://github.com/user-attachments/assets/abd6ff1c-52b6-4770-b894-7dd5e5dbaf11)

**Reply by m0sth8** [reply]
Thank you. Haven't noticed it. I've just transferred repo. Thank you for taking care of it.

**Reply by dearchap** [reply]
Merged my changes from urfave/sflags-bak into urfave/sflags and deleted the backup repo. !!! Thanks @m0sth8
